gym_wrapper/base_gym_env.py

```diff
@@ -47,103 +61,93 @@ def __init__(
-            )
+        # check for valid max_buttons_pressed
+        if max_buttons_pressed > self.num_binary_buttons > 0:
+            warnings.warn(
+                f"max_buttons_pressed={max_buttons_pressed} "
+                f"> number of binary buttons defined={self.num_binary_buttons}. "
+                f"Clipping max_buttons_pressed to {self.num_binary_buttons}.")
+            max_buttons_pressed = self.num_binary_buttons
+        elif max_buttons_pressed < 0:
+            raise RuntimeError(f"max_buttons_pressed={max_buttons_pressed} < 0. Should be >= 0. ")
 
-        if self.automap:
-            spaces["automap"] = gym.spaces.Box(
-                0,
-                255,
-                (
-                    self.game.get_screen_height(),
-                    self.game.get_screen_width(),
-                    # "automap" buffer uses same number of channels
-                    # as the main screen buffer
-                    3,
-                ),
-                dtype=np.uint8,
-            )
+        # specify action space(s)
+        self.max_buttons_pressed = max_buttons_pressed
+        self.action_space = self.__get_action_space(continuous_action_space)
 
-        self.num_game_variables = self.game.get_available_game_variables_size()
-        if self.num_game_variables > 0:
-            spaces["gamevariables"] = gym.spaces.Box(
-                np.finfo(np.float32).min,
-                np.finfo(np.float32).max,
-                (self.num_game_variables,),
-                dtype=np.float32
-            )
+        # specify observation space(s)
+        self.observation_space = self.__get_observation_space()
 
-        self.observation_space = gym.spaces.Dict(spaces)
+        self.game.init()
 
     def step(self, action):
         assert self.action_space.contains(action), f"{action!r} ({type(action)}) invalid"
         assert self.state is not None, "Call `reset` before using `step` method."
 
-        # convert action to vizdoom action space (one hot)
-        act = [0 for _ in range(self.action_space.n)]
-        act[action] = 1
-
-        reward = self.game.make_action(act, self.frame_skip)
+        env_action = self.__build_env_action(action)
+        reward = self.game.make_action(env_action, self.frame_skip)
         self.state = self.game.get_state()
         done = self.game.is_episode_finished()
 
         return self.__collect_observations(), reward, done, {}
 
+    def __parse_binary_buttons(self, env_action, agent_action):
+        if self.num_binary_buttons != 0:
+            if self.num_delta_buttons != 0:
+                agent_action = agent_action["binary"]
+
+            if isinstance(agent_action, int):
+                # agent action maps to numerically equivalent environment action (i.e. bit string is the env action)
+                agent_action = self.button_map[agent_action]
+                agent_action = np.array([i for i in range(agent_action.bit_length()) if agent_action & (1 << i)])
```

Comment:
Sidenote: this logic might be difficult to debug if errors arise. You could use something [like this](https://github.com/Miffyli/rl-action-space-shaping/blob/master/envs/vizdoom_environment.py#L151-L171) to create a list of valid actions ("binary strings") and then map the discrete action number to action by indexing that list.